@here/harp-mapview/lib/TextureLoader.ts

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (C) 2020 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import * as THREE from "three";
+
+export interface RequestHeaders {
+    [field: string]: string;
+}
+
+/**
+ * A texture loader that also supports request headers(e.g. Authorization)
+ */
+export class TextureLoader extends THREE.TextureLoader {
+    requestHeaders?: RequestHeaders;
+
+    constructor(manager?: THREE.LoadingManager) {
+        super(manager);
+        // empty assignment required to support CORS
+        this.setCrossOrigin("");
+    }
+
+    setRequestHeaders(requestHeaders: RequestHeaders | undefined): void {
+        this.requestHeaders = requestHeaders;
+    }
+
+    load(
+        url: string,
+        onLoad?: (texture: THREE.Texture) => void,
+        onProgress?: (event: ProgressEvent) => void,
+        onError?: (event: ErrorEvent) => void
+    ): THREE.Texture {
+        // Use normal texture loader if no request header is set
+        if (this.requestHeaders === undefined) {
+            return super.load(url, onLoad, onProgress, onError);
+        }
+
+        if (this.path !== undefined) {
+            url = this.path + url;
+        }
+        url = this.manager.resolveURL(url);
+
+        // Load image with fetch API if request header is set
+        const texture = new THREE.Texture();
+
+        this.manager.itemStart(url);
+        fetch(url, {
+            headers: this.requestHeaders,
+            mode: this.crossOrigin !== undefined ? "cors" : "no-cors"
+        })
+            .then(response => {
+                return response.blob();
+            })
+            .then(blob => {
+                return this.loadImageFromBlob(blob);
```

Comment:
I don't think that we need to copy almost whole original TextureLoader here.

When you have blob, you can `super.load(URL.createObjectUrl(blob), ...).

See https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL

This should have exactly same complexity as code you've written here.

Reply:
I don't think this is possible if I want to keep the interface of `TextureLoader.load()` compatible. The function creates a texture that is immediately returned while starting the async loading of the image. But in order to call `super.load(URL.createObjectUrl(blob), ...)` the blob needs to be already loaded. The texture created by `super.load()` can't be returned immediately because the `fetch()` request needs to be completed before `super.load()` can be called. Additionally, the texture format(RGB vs RGBA) will not be detected correctly because the file format ending is only in the original url but not the object url.

Reply:
You're right. Sorry for forcing you to write it, it's obvious.

Then why stick with "old", callback-era interface of TextureLoader. Do we really need to be compatible with original loader if we already reimplemented.

Why we can't have "our" texture loader with simple and robust promise-based API ?
``
const texture = await ourTextureLoader.load(...) // Proimise<texture>
```

Reply:
Answering potential question.

I know that original 3JS API returns empty texture synchronously, so we can use it, but i'm not sure if there is any use for texture that is not loaded, so API returning `Promise<Texture>` should be ok for us.

Reply:
I refactored the new Texture loaded to use a Promise based API.